packages/css/src/types.ts

```diff
@@ -465,24 +464,37 @@ export interface VariantProperty {
    *
    * @see https://styled-system.com/variants
    */
-  variant: string
+  variant?: string
 }
 
-export interface UseThemeFunction {
-  (theme: any): Exclude<ThemeUIStyleObject, UseThemeFunction>
+export interface ThemeDerivedStyles {
+  (theme: Theme): Exclude<ThemeUIStyleObject, ThemeDerivedStyles>
```

Comment:
I added `Theme` type to the argument here.

This is a breaking change for TypeScript ppl who added extra properties to their theme object and use them in functions.

- [ ] I'll draft a TypeScript section in the docs, and describe how to add additional properties to the Theme type.

Reply:
> I have no idea how common of a use case that is

It actually affects half of my projects using Theme UI. 😢

This is also breaking `(theme: MyExactTheme) => ...` usage, where components depend on exact shape of the theme. This makes sense a lot of sense for blogs, but it's a bug in apps where the theme can be customized by the user.

I have mixed feelings about this change.
Possibly, we could add a generic parameter and describe derived styles as `<T extends Theme = Theme>(theme: T) => ThemeUICSSObject` to allow this 🤔

```
sx={{
  bg: (t: MyTheme) => colorMode === 'dark'
    ? transparentize(t.colors.modes.dark.primary, 0.2)
    : darken(t.styles.form.backgroundColor, 0.6)
}}
```

Reply:
> Possibly, we could add a generic parameter and describe derived styles as `<T extends Theme = Theme>(theme: T) => ThemeUICSSObject` to allow this

That sounds pretty good...?

Reply:
> That sounds pretty good...?

Too good :/ It doesn't work.

Reply:
Of course. 🤦‍♂️